File: 03-herencia/java/Auto.java
/*
 * Auto.java - CLASE HIJA
 *
 * "extends Vehiculo" se lee: un Auto ES UN Vehiculo.
 * Si no puedes decir "ES UN" en voz alta, probablemente no es herencia.
 *
 * Fijate que este archivo NO importa nada: Vehiculo esta en la misma
 * carpeta, o sea en el mismo paquete.
 */
public class Auto extends Vehiculo {

    private int puertas;   // atributo que el padre NO tiene

    public Auto(String modelo, int anio, int puertas) {
        super(modelo, anio);      // llama al constructor del padre. SIEMPRE primero.
        this.puertas = puertas;
    }

    @Override                     // avisa al compilador: estoy reemplazando el del padre
    public void describir() {
        System.out.println("Auto " + this.modelo + " (" + this.anio + "), "
                           + this.puertas + " puertas");
    }

    public void abrirMaletero() {        // metodo exclusivo de Auto
        System.out.println(this.modelo + ": maletero abierto");
    }
}

File: 03-herencia/python/auto.py
"""
auto.py - CLASE HIJA

  class Auto extends Vehiculo   ->   class Auto(Vehiculo):

A diferencia de Java, aqui SI hay que importar al padre aunque este
en la misma carpeta.
"""

from vehiculo import Vehiculo


class Auto(Vehiculo):

    def __init__(self, modelo, anio, puertas):
        super().__init__(modelo, anio)     # llama al constructor del padre
        self._puertas = puertas

    def describir(self):                   # sobrescribe el del padre
        print(f"Auto {self._modelo} ({self._anio}), {self._puertas} puertas")

    def abrir_maletero(self):              # metodo exclusivo de Auto
        print(f"{self._modelo}: maletero abierto")
